addons/project_timesheet_holidays/models/hr_holidays.py

```diff
@@ -151,3 +151,14 @@ def _force_cancel(self, *args, **kwargs):
         timesheets = self.sudo().timesheet_ids
         timesheets.holiday_id = False
         timesheets.unlink()
+
+    def write(self, vals):
+        res = super().write(vals)
+        # reevaluate timesheets after the leaves are wrote in order to remove empty timesheets
+        empty_timesheets = self.env['account.analytic.line'].browse()
+        for leave in self:
+            if leave.number_of_days == 0 and leave.timesheet_ids and leave.timesheet_ids.holiday_id:
+                leave.timesheet_ids.holiday_id = False
+                empty_timesheets |= leave.timesheet_ids
+        empty_timesheets.unlink()
```

Comment:
The union operator will each time do a browse, since we only need the recordset at the end of the for loop, we could just do the browse at the end of the for loop to only browse one time. 🙂 

```suggestion
        timesheet_ids_to_remove = []
        for leave in self:
            if leave.number_of_days == 0 and leave.timesheet_ids and leave.timesheet_ids.holiday_id:
                leave.timesheet_ids.holiday_id = False
                timesheet_ids_to_remove.extends(leave.timesheet_ids)
        self.env['account.analytic.line'].browse(set(timesheet_ids_to_remove)).sudo().unlink()
```

Moreover, we will have to do it in sudo, since the current user could have no access to timesheets. 🙂

Reply:
Thank you for the review! I am glad to learn about the issue with browse here :)